### DISCUSSION TITLE
Can you make a repeat note/chord function?

### DISCUSSION THREAD

**Discussion by Sy256** [helge17/tuxguitar#24] [General]
Hello I love tuxguitar...well, I loved it more in its old version the new one seems to be lacking some things the old one had but only the new one works on my new mac (M1) sooooo yeah....the ONE thing I'd love is to have a repeat note/chord button that'll repeat the highlighted note/chord for the same duration when you press a key. Musescore has this, but I don't want to use musescore. For one, musescore lacks the move the beat function. Tuxguitar is great, I like it more than guitar pro for my purposes, but I miss this feature.

I also don't really know how github works in general, how do I download the new one you guys are maintaining here? 
Thanks.

If you can't make this / its not easy, would you please advise how you personally do it when having to enter 6-string chords multiple times in the same measure? Am I missing a time-saving step? Even the ability to copy/paste within measures and not just entire measures would cover this.

**Comment by guiv42** [comment]
Hi
No idea of the complexity of such a feature. How does it look like in MuseScore?

> how do I download the new one you guys are maintaining here?

You can download "snapshot" versions (i.e. work in progress) here: https://github.com/helge17/tuxguitar/releases
Just select the most recent date (2023-05-01 as of today), then choose the version adapted to your platform

**Reply by Sy256** [reply]
Thanks! I will try that. Musescore's repeat note thing looks like this

I couldn't find an example from musescore so I got one from sibelius, but keep in mind it doesn't have to be this advanced https://youtu.be/eVXzBtqYHLM?t=76

I just want to be able to hit a key and have it repeat the one note/chord that is highlighted. Perhaps you could use the already existing chord library as a starting point? Like...save the highlighted chord, recall and add to the tab, delete from the library, so fast that we don't see the dialog for the library? I don't know I'm just spitballing here lol, I don't know whats possible with programming but it seems like a macro possibly.

**Reply by guiv42** [reply]
Hi
There is a very interesting "click & drag  select" feature in the (now inactive) "TuxGuitar2.0beta" branch. Selection really looks like what you are showing there. Not strictly a Repeat feature, but at least you could select-copy-paste within 1 measure.
If I can get it to work, I guess the most complex part would be done. "Select-Repeat" is nothing more than "Select-(copy-moveRight-paste)", it would probably be easy.
That being said, integration of this feature is not trivial, but at least I'll have a try.
If possible, help would be welcome:
- I have a few questions about user interface (see #29), could you please give your opinion?
- Could you contribute to testing? This would require you to build from source, as I cannot build for MacOS environment

**Reply by Sy256** [reply]
Hi! That sounds awesome and I'd be up to beta test (I know nothing about programming but I can test it) and I'd love to try the click and drag buttttt none of the releases are able to be opened on my computer you got any ideas? I am on M1 Macbook Air Monterey. The official tuxguitar version macosx-cocoa-64 1.5.6 opens but none of your release do. I get the message 'do you want to open this application unknown develop etc etc' and say yes and then just keep clicking it and nothing opens.

If the click drag works I don't even need a repeat function anymore but it'd still be nice to have if not hard to do.

**Reply by guiv42** [reply]
Hum... this mac issue is really annoying. I have zero knowledge in Apple's ecosystem, and nothing available to test :-(
But we definitely need to tackle that issue. Since this is a completely independent problem, could you please create a dedicated [issue](https://github.com/helge17/tuxguitar/issues), with all details? Tested release(s), configuration of the machine, and a screenshot of the error message would be useful data

**Reply by Sy256** [reply]
Ok thanks I just made an issue on it as best I could.

Also I've been trying to install the tuxguitar 2.0 beta from a different github homebrew but don't know how and keep getting errors. Are you part of this project since you'd mentioned it? https://github.com/pterodactylus42/tuxguitar-2.0beta


The error I got is here : https://pastebin.com/7yJr5tSF

After downloading the separate SWT folder and trying to install that...I guess. I'm just doing what people are telling me I have no idea how this works.

**Reply by guiv42** [reply]
If you want to have a try at the 2.0beta branch you shouldn't need to build from source. helge17 has published builds in his repo: https://github.com/helge17/tuxguitar-2.0beta/releases
However you will most probably have the same execution issue in your environment.
Thanks for creating the issue, hope someone will be able to help.
And no, I haven't been part of any past project. I've been a TG user since quite a long time, but never involved in dev. Just arrived a few weeks ago in this activity, and I'm currently  discovering all of this

**Comment by guiv42** [comment]
@Sy256 : could you please have a look at discussion #29?
There are a few open questions about user interface, and a link to a first version for beta-testing (not everything implemented yet)

**Comment by guiv42** [comment]
@Sy256 this is just a teaser, I hope you will like it ;-)

https://github.com/helge17/tuxguitar/assets/129443524/79e2696e-0d1e-4304-a41a-e61a37f53e3e

**Comment by guiv42** [comment]
@Sy256 : select/cut/copy/paste and also repeat features are now available in my fork, not yet merged here. If you want to participate to beta-testing you're welcome! I have provided [builds for Linux & Windows](https://github.com/guiv42/tuxguitar/releases), but yet I cannot build for mac.

@helge17 : what do you think is the best way to go forward? I can see 2 options:
1. merge my branch as it is, and proceed to fine-tuning and eventually bug fixes here
2. wait a bit for tests by @Sy256 before merging. In this eventuality could you please provide a binary build for mac?

What is your preference?

**Reply by helge17** [reply]
1. is better i think!

**Comment by guiv42** [comment]
Thanks @helge17 for merging!
So I suggest to close discussion #29, and to manage fine-tuning and bug fixes through "issues".
Since the evolution is quite important, I think we need to update [CHANGES](https://github.com/helge17/tuxguitar/blob/master/CHANGES) file. Somehow we also need to credit the authors of this "click & drag select" feature. I will send a new pull request for that.
Possibly it could be also interesting to update related "help" section?
Note: "Repeat" string also needs to be translated in [share/lang](https://github.com/helge17/tuxguitar/tree/master/TuxGuitar/share/lang)

**Comment by guiv42** [comment]
@Sy256 : any feedback?
Now that the feature is merged, can we close this discussion?

**Reply by Sy256** [reply]
Hi! Sorry I was away and didn't see notifications that there were new messages here. That video of the demo looks great! Thank you! I am downloading the latest beta version to try but my internet is slow, I'll report back once its done. I came up with another wishlist of features I'll post separately

**Reply by Sy256** [reply]
It is working amazing great job guys!

**Comment by Sy256** [comment]
Hello I actually do have a note to reopen this and if the answer's no then I'll close it. I noticed the repeat function overrides whatever is on the clipboard, is there a way around that, can we have 2 clipboards one for copy/paste and one for repeating? I'm loving the function just wondering if this is possible. Thanks

**Reply by guiv42** [reply]
After a short analysis, it seems such a feature is quite complex to implement

**Comment by guiv42** [comment]
Just a suggestion: close this discussion now?
